src/ui/public/visualize/visualization.js

```diff
@@ -115,50 +116,64 @@ uiModules
 
         $scope.vis.initialized = true;
 
-        const renderFunction = _.debounce(() => {
-          const container = getVisContainer();
-          if (!container) return;
-          $scope.vis.size = [container.width(), container.height()];
-          const status = getUpdateStatus($scope);
-          visualization.render($scope.visData, status)
-            .then(() => {
-            // renderComplete
-              $scope.$emit('renderComplete');
-              $el.trigger('renderComplete');
+        const dispatchCustomEvent = (name) => {
+          // we're using the native events so that we aren't tied to the jQuery custom events,
+          // otherwise we have to use jQuery(element).on(...) because jQuery's events sit on top
+          // of the native events per https://github.com/jquery/jquery/issues/2476
+          $el[0].dispatchEvent(new CustomEvent(name, { bubbles: true }));
+        };
+
+        const render$ = Observable.create(observer => {
+          $scope.$on('render', () => {
+            observer.next({
+              vis: $scope.vis,
+              visData: $scope.visData,
+              container: getVisContainer(),
             });
-          $scope.$apply();
-        }, 100);
-
-        $scope.$on('render', () => {
-          if (!$scope.vis || !$scope.vis.initialized || ($scope.vis.type.requiresSearch && !$scope.visData)) {
-            return;
-          }
-          $scope.addLegend = $scope.vis.params.addLegend;
-          $scope.vis.refreshLegend++;
-          $timeout(renderFunction);
+          });
         });
 
+        const success$ = render$
+          .do(() => {
+            dispatchCustomEvent('renderStart');
+          })
+          .filter(({ vis, visData, container }) => vis && vis.initialized && container && (!vis.type.requiresSearch || visData))
+          .do(({ vis }) => {
+            $scope.addLegend = vis.params.addLegend;
+            vis.refreshLegend++;
+          })
+          .debounceTime(100)
+          .switchMap(async ({ vis, visData, container }) => {
+            vis.size = [container.width(), container.height()];
+            const status = getUpdateStatus($scope);
+            const renderPromise = visualization.render(visData, status);
+            $scope.$apply();
+            return renderPromise;
+          });
+
+        const requestError$ = render$.filter(({ vis }) => vis.requestError);
+
+        const renderSubscription = Observable.merge(success$, requestError$)
+          .subscribe(() => {
+            window.requestAnimationFrame(() => {
```

Comment:
why are we using `requestAnimationFrame` here instead of `setTimeout` as usually?

Reply:
apart from the use of requestAnimationFrame (no preference wrt. setTimeout)

Is there any reason we are unhooking these events from the call-stack? I removed it, and things seemed to work fine.

Reply:
We shouldn't need to use setTimeout or requestAnimationFrame, left over from trial and error to get this firing correctly.